File: packages/gui/src/pages/settings/platformMapRows.ts
export type PlatformMapSource = "default" | "override" | "identity";

export interface PlatformMapRow {
  rommSlug: string;
  esdeFolder: string;
  source: PlatformMapSource;
}

export function defaultEsdeFolder(
  rommSlug: string,
  bundled: Record<string, string>,
): string {
  return bundled[rommSlug] ?? rommSlug;
}

export function rowSource(
  rommSlug: string,
  esdeFolder: string,
  bundled: Record<string, string>,
): PlatformMapSource {
  const fallback = defaultEsdeFolder(rommSlug, bundled);
  if (esdeFolder === fallback) {
    return bundled[rommSlug] !== undefined ? "default" : "identity";
  }
  return "override";
}

export function buildPlatformMapRows(
  bundled: Record<string, string>,
  overrides: Record<string, string>,
): PlatformMapRow[] {
  const slugs = new Set([
    ...Object.keys(bundled),
    ...Object.keys(overrides),
  ]);
  return [...slugs].sort((a, b) => a.localeCompare(b)).map((rommSlug) => {
    const esdeFolder =
      overrides[rommSlug] ?? defaultEsdeFolder(rommSlug, bundled);
    return {
      rommSlug,
      esdeFolder,
      source: rowSource(rommSlug, esdeFolder, bundled),
    };
  });
}

export function overridesFromRows(
  rows: PlatformMapRow[],
  bundled: Record<string, string>,
): Record<string, string> {
  const out: Record<string, string> = {};
  for (const row of rows) {
    const fallback = defaultEsdeFolder(row.rommSlug, bundled);
    if (row.esdeFolder !== fallback) {
      out[row.rommSlug] = row.esdeFolder;
    }
  }
  return out;
}

export const PLATFORM_MAP_SOURCE_LABEL: Record<PlatformMapSource, string> = {
  default: "Default",
  override: "Override",
  identity: "Identity",
};

File: src/shared/src/commonMain/kotlin/dev/rommdeck/shared/platform/PlatformMapRows.kt
package dev.rommdeck.shared.platform

enum class PlatformMapSource {
    Default,
    Override,
    Identity,
}

data class PlatformMapRow(
    val rommSlug: String,
    val esdeFolder: String,
    val source: PlatformMapSource,
)

fun defaultEsdeFolder(rommSlug: String, bundled: Map<String, String>): String =
    bundled[rommSlug] ?: rommSlug

fun rowSource(
    rommSlug: String,
    esdeFolder: String,
    bundled: Map<String, String>,
): PlatformMapSource {
    val fallback = defaultEsdeFolder(rommSlug, bundled)
    if (esdeFolder == fallback) {
        return if (bundled.containsKey(rommSlug)) PlatformMapSource.Default else PlatformMapSource.Identity
    }
    return PlatformMapSource.Override
}

fun buildPlatformMapRows(
    bundled: Map<String, String>,
    overrides: Map<String, String>,
): List<PlatformMapRow> {
    val slugs = (bundled.keys + overrides.keys).toSortedSet()
    return slugs.map { rommSlug ->
        val esdeFolder = overrides[rommSlug] ?: defaultEsdeFolder(rommSlug, bundled)
        PlatformMapRow(
            rommSlug = rommSlug,
            esdeFolder = esdeFolder,
            source = rowSource(rommSlug, esdeFolder, bundled),
        )
    }
}

fun overridesFromRows(
    rows: List<PlatformMapRow>,
    bundled: Map<String, String>,
): Map<String, String> {
    val out = linkedMapOf<String, String>()
    for (row in rows) {
        val fallback = defaultEsdeFolder(row.rommSlug, bundled)
        if (row.esdeFolder != fallback) {
            out[row.rommSlug] = row.esdeFolder
        }
    }
    return out
}

fun platformMapSourceLabel(source: PlatformMapSource): String = when (source) {
    PlatformMapSource.Default -> "Default"
    PlatformMapSource.Override -> "Override"
    PlatformMapSource.Identity -> "Identity"
}
